services/api/src/unified_graphics/diag.py

```diff
@@ -69,120 +68,63 @@ def to_geojson(self):
-    """Opens a diag file stored in S3
-
-    Assumes AWS S3, and grabs credentials via Boto3 defaults.
-
-    If FLASK_ENDPOINT_URL is set in the environment, it will attempt to access
-    an S3-like server at the URL provided. This is mostly useful for testing.
-    """
-    try:
-        endpoint_url = current_app.config["ENDPOINT_URL"]
-    except KeyError:
-        endpoint_url = None
-    cache = current_app.config["LOCAL_CACHE_PATH"]
-    current_app.logger.info(f"Endpoint url set to: {endpoint_url}")
-    diag_file = diag_uri + filename  # Path doesn't support file URIs
-
-    # xarray.open_dataset can't open netcdf datasets natively - only
-    # Zarr datasets. We'll use fsspec to cache the file locally so xarray
-    # can access it from disk.
-    try:
-        # TODO - if statement is a workaround
-        # TODO - for: https://github.com/fsspec/s3fs/issues/400
-        if endpoint_url:
-            with fsspec.open(
-                f"simplecache::{diag_file}",
-                s3={"anon": False, "client_kwargs": {"endpoint_url": endpoint_url}},
-                simplecache={"cache_storage": cache},
-            ) as f:
-                ds_contents = xr.open_dataset(f.name)
-        else:
-            with fsspec.open(
-                f"simplecache::{diag_file}",
-                s3={"anon": False},
-                simplecache={"cache_storage": cache},
-            ) as f:
-                ds_contents = xr.open_dataset(f.name)
-    except FileNotFoundError as e:
-        current_app.logger.error(
-            f"The following bucket or key don't exist: '{str(diag_file)}'",
-            exc_info=True,
-        )
-        raise e
-    except PermissionError as e:
-        current_app.logger.error(
-            f"Permissions issue accessing: '{str(diag_file)}'", exc_info=True
-        )
-        raise e
-    except ValueError as e:
-        current_app.logger.error(
-            f"Unknown file type: '{str(diag_file)}'", exc_info=True
-        )
-        raise e
-    return ds_contents
+def get_store(url: str) -> Union[str, S3Map]:
+    result = urlparse(url)
+    if result.scheme in ["", "file"]:
+        return result.path
+
+    if result.scheme != "s3":
+        raise ValueError(f"Unsupported protocol '{result.scheme}' for URI: '{url}'")
+
+    region = os.environ.get("AWS_REGION", "us-east-1")
+    s3 = S3FileSystem(
+        key=os.environ.get("AWS_ACCESS_KEY_ID"),
+        secret=os.environ.get("AWS_SECRET_ACCESS_KEY"),
+        token=os.environ.get("AWS_SESSION_TOKEN"),
+        client_kwargs={"region_name": region},
+    )
+
+    return S3Map(root=f"{result.netloc}{result.path}", s3=s3, check=False)
+
+
+def open_diagnostic(
+    variable: Variable, initialization_time: str, loop: MinimLoop
+) -> xr.Dataset:
+    store = get_store(current_app.config["DIAG_ZARR"])
+    group = f"/{variable.value}/{initialization_time}/{loop.value}"
+    return xr.open_zarr(store, group=group)
```

Comment:
👏 Using `urlparse` & `s3fs` here in place of `fsspec` really does remove a good amount of complexity. Nicely done.

Reply:
Yeah, it’s one of the advantages of Zarr. Pretty sure this wouldn’t work with NetCDF, although I’ve never tried.